published/external/xdp/framelayout.h

```diff
@@ -0,0 +1,62 @@
+//
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+//
+
+#pragma once
+
+EXTERN_C_START
+
+#pragma warning(push)
+#pragma warning(default:4820) // warn if the compiler inserted padding
+#pragma warning(disable:4201) // nonstandard extension used: nameless struct/union
+#pragma warning(disable:4214) // nonstandard extension used: bit field types other than int
+
+typedef enum _XDP_FRAME_LAYER2_TYPE {
+    XdpFrameLayer2TypeUnspecified,
+    XdpFrameLayer2TypeNull,
+    XdpFrameLayer2TypeEthernet,
+} XDP_FRAME_LAYER2_TYPE;
+
+typedef enum _XDP_FRAME_LAYER3_TYPE {
+    XdpFrameLayer3TypeUnspecified,
+    XdpFrameLayer3TypeIPv4UnspecifiedOptions,
+    XdpFrameLayer3TypeIPv4WithOptions,
+    XdpFrameLayer3TypeIPv4NoOptions,
+    XdpFrameLayer3TypeIPv6UnspecifiedExtensions,
+    XdpFrameLayer3TypeIPv6WithExtensions,
+    XdpFrameLayer3TypeIPv6NoExtensions,
+} XDP_FRAME_LAYER3_TYPE;
+
+typedef enum _XDP_FRAME_LAYER4_TYPE {
+    XdpFrameLayer4TypeUnspecified,
+    XdpFrameLayer4TypeTcp,
+    XdpFrameLayer4TypeUdp,
+    XdpFrameLayer4TypeIPFragment,
+    XdpFrameLayer4TypeIPNotFragment,
+} XDP_FRAME_LAYER4_TYPE;
+
+#include <pshpack1.h>
```

Comment:
Why use includes for pack? I've never seen that before.

Reply:
It was recommended years ago by Jeffrey, and it is a tidy one-liner. I don't have a particularly strong opinion, other than to say that this code was just copied from offload.h. If I had written this from scratch, I probably would have forgotten these headers even exist.

Perhaps one benefits is that phspack1.h abstracts away compiler-specific push/pack pragmas.

Reply:
Ok. I'm fine with it. Just never seen it before.